### PR TITLE
modify the perf test for kernel sobel

### PR DESCRIPTION
**Pull request by lionkunonly**
Modify the perf test for kernel Sobel by referring the native perf test.

### CODE REVIEW

File: modules/js/perf/perf_imgproc/perf_sobel.js [hunk 1]
```diff
@@ -25,43 +25,40 @@ function perf() {
...
-                }
+                let dxdy = this.params.dxdy;
+                let ksize = this.params.ksize;
                 
-                if (direction == "dx") {dx = 1;}
-                else {dy = 1;}
+                let src = new cv.Mat(size[1], size[0], cv.CV_8U);
+                let dst = new cv.Mat(size[1], size[0], ddepth);
+                let dx = parseInt(dxdy[1]);
+                let dy = parseInt(dxdy[dxdy.length-2]);
```
Comment (terfendail) [inline modules/js/perf/perf_imgproc/perf_sobel.js]: Is evaluation `dxdy.length-2` really necessary? I suppose the result is always equal to zero.

File: modules/js/perf/perf_imgproc/perf_sobel.js [hunk 2]
```diff
@@ -88,44 +96,48 @@ function perf() {
         let suite = new Benchmark.Suite;
         totalCaseNum = 0;
         currentCaseId = 0;
+        let params = "";
+        let paramObjs = [];
+        paramObjs.push({name:"size", value:"", reg:[""], index:0});
+        paramObjs.push({name:"ddepth", value:"", reg:["/CV\_[0-9]+[A-z]+[A-z]+[0-9]/g"], index:1});
```
Comment (terfendail) [inline modules/js/perf/perf_imgproc/perf_sobel.js]: Regexp could be a bit simpler here. First variable letter is from set `[FSUfsu]` while second is always `C`. Also generally there could be more than 9 channels, however I've never seen a test with such input.

File: modules/js/perf/perf_imgproc/perf_sobel.js [hunk 3]
```diff
@@ -88,44 +96,48 @@ function perf() {
...
+        paramObjs.push({name:"dxdy", value:"", reg:["/\\([0-2]+,[0-2]\\)/"], index:2});
```
Comment (terfendail) [inline modules/js/perf/perf_imgproc/perf_sobel.js]: First part of the parameter is also single digit

File: modules/js/perf/perf_imgproc/perf_sobel.js (hunk 1 above)
Comment (lionkunonly) [reply modules/js/perf/perf_imgproc/perf_sobel.js]: I think it is necessary. The parameter `dxdy` represents the tuple of the `dx` and `dy`.  In the native perf test,  there are some options of the tuple of `dx` and `dy`, like `(0, 1)`, `(1, 0)`, `(1, 1)`, `(2, 0)` and so on.

File: modules/js/perf/perf_imgproc/perf_sobel.js (hunk 2 above)
Comment (lionkunonly) [reply modules/js/perf/perf_imgproc/perf_sobel.js]: I have simplified it. Please review it.

File: modules/js/perf/perf_imgproc/perf_sobel.js (hunk 3 above)
Comment (lionkunonly) [reply modules/js/perf/perf_imgproc/perf_sobel.js]: I have simplified it. Please review it.

File: modules/js/perf/perf_imgproc/perf_sobel.js (hunk 1 above)
Comment (terfendail) [reply modules/js/perf/perf_imgproc/perf_sobel.js]: Looks like the tuple length is constant and always equals 2. Anyway if the tuple length could variate do we always want second from the end value?

Comment (lionkunonly) [reply modules/js/perf/perf_imgproc/perf_sobel.js]: Ok, the `dxdy.length-2` is suitable for my implementation. In my implementation, I set the `dxdy` as a string type data with a format like `(0,1)`, `(1,0)`, `(1,1)` and so on. So It is ok to replace `dxdy.length-2` with `3`. I will change it in the next commit.

**Comment by lionkunonly** [comment]
@terfendail  Hi, Vitaly. I have modified this PR based on your comments. If it is ok, I think this PR can be merged.

File: modules/js/perf/perf_imgproc/perf_sobel.js (hunk 1 above)
Comment (terfendail) [reply modules/js/perf/perf_imgproc/perf_sobel.js]: I get the idea. Sorry I've messed the length of the string with amount of the tuple elements. My bad.

Anyway 3 should work better for persons not familiar with javascript like me.

**Review by terfendail — APPROVED** [review]

**Comment by terfendail** [comment]
I think we should wait for input from @huningxin if the merge isn't a development stopper.

**Review by huningxin — APPROVED** [review]
Looks good.

**Comment by huningxin** [comment]
I am going to merge this PR. @lionkunonly , please incorporate this change into https://github.com/opencv/opencv/pull/18068.

**Comment by lionkunonly** [comment]
@huningxin This change has been merged in the https://github.com/opencv/opencv/pull/18068 some days ago.

**Pull request merged**